Gravatar/MruImageCache.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Gravatar
+{
+    /// <summary>
+    /// Caches most-recently-used images.
+    /// </summary>
+    /// <remarks>
+    /// Decorates an inner cache, delegating to it as needed.
+    /// <para />
+    /// If an image is available in memory, the inner cache can be bypassed.
+    /// </remarks>
+    public sealed class MruImageCache : IImageCache
+    {
+        private const int CleanAtSize = 30;
+        private const int CleanToSize = 25;
+
+        private readonly Dictionary<string, Entry> _entryByFileName = new Dictionary<string, Entry>();
+        private readonly IImageCache _inner;
+
+        public MruImageCache(IImageCache inner)
+        {
+            _inner = inner;
+        }
+
+        event EventHandler IImageCache.Invalidated
+        {
+            add => _inner.Invalidated += value;
+            remove => _inner.Invalidated -= value;
+        }
+
+        void IImageCache.AddImage(string imageFileName, Image image)
```

Comment:
all interface-based methods (which makes them public in essence) must validate parameters as they (almost) all are NRE prone. E.g. if `DeleteImageAsync(null)` will throw because `Dictionary.Remove(null)` throws...

Reply:
Can do.

I'll annotate the API and make behaviour consistent across implementations. For example, `DirectoryImageCache.AddImage` just returns if `imageFileName` is `null`.

Reply:
Annotations added. Validation and tests updated.